Finished ion-compare basic mode

cheqianh opened the pull request:
### Description:
This pr finished ion-compare basic mode and based on https://github.com/cheqianh/ion-java/pull/2
I will make another pull request for equivs/non-equivs/equivs-timeline modes separately.

The compare function convert ion-stream or event-stream files to event structure lists and compare the lists.

### Note:

1. I worked on compare part and process part in parallel, so I didn't refactor some functions that show up in both files. I'll refactor them after done ion-compare and ion-process. I've already make a file for refactor functions which is called IonJavaTestUtil.java. (or other name)

2. The signature of compare is: 
`compare(CompareContext compareContext,  boolean writeReport, int startI, int endI, int startJ, int endJ)  `
 **compareContext** 
CompareContext stores eventStream(ArrayList\<Event\>),
 **writeReport** 
WriteReport is for deciding weather writing content to the report or not. There are 2 ways to use compare function: one is Comparing eventStream and write a comparisonReport, and the second way is using it as a helper function that simply Compare two eventStreams. So writeReport is false if we just want to compare two random eventStreams.
**int startI, int endI** 
These 2 numbers represent the range of the left hand eventStream that going to be compared.
**int startJ, int endJ** 
These 2 numbers represent the range of the right hand eventStream that going to be compared.
**For example:**
StartI, endI are 1,3. startJ and endJ are 2,4. It means we are going to compare sublist of the left eventStream in index 1-3 to sublist of the right hand eventStream in index 2-4.

&nbsp;&nbsp;isSameStruct method which compares 2 structs has similar arguments.

3. To compare if 2 structs are same. I scan the first struct and Store each field in a hashMap(key is fieldName and value is a list of pairs which each pair represents the index of the range that this field shows in the eventStream), then I go through the second struct and check if each element shows up in the hashMap: if it's found then remove this data from the map, otherwise return false.

### Test
1. Embedde stream
2. List, sexp and other scalar value
3. Struct, nested struct
4. Mixed of above

cheqianh committed:
compare set up

finish get events

finish embeddedstream

finish write output

finish parse container

working on map

restructure

almost finish continaer

fix unordered bug

finished contianer check

ready for improve algorithm

remove unnecessary variables

reduce space complexity

cleaned some useless code

remove useless comparecontext variables

fix bugs and find embedded bug

fix embedded bug

check style

work on equivs/non-equivs

refactor and fix embedded bug

finish refactor

ready for cr

cheqianh committed:
little change

tgregg commented on test/tools/cli/ComparisonType.java:
```diff
@@ -0,0 +1,8 @@
+    BASIC,
+    EQUIVS,
+    NON_EQUIVS,
+    EQUIVSS_TIMELINE
```
EQUIVSS_TIMELINE -> EQUIVS_TIMELINE

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/CompareContext.java:
```diff
@@ -0,0 +1,88 @@
+    String compareToPath;
+    int fileEventIndex;
+    int compareToFileEventIndex;
+    ArrayList<Event> eventStreamFirst;
+    ArrayList<Event> eventStreamSecond;
```
These can probably just be declared as `List<Event>`--it's not important that the underlying implementation is `ArrayList`.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,771 @@
+    private static final IonTextWriterBuilder ION_TEXT_WRITER_BUILDER = IonTextWriterBuilder.standard();
+    private static final String EMBEDDED_STREAM_ANNOTATION = "embedded_documents";
+
+    public static void main(final String[] args) {
```
Both `process` and `compare` should be accessible via commands to the same executable. Will that work?

cheqianh replied:
Sure, I'm going to create a new file that includes process and compare.

Before I start restructure files, can I double check two things:
1. We can't call compare and process in one command right?
2. Do you think it's a good idea to make a new file 'IonCli' that parses the command and call the corresponding process/compare functions?

tgregg replied:
1. Correct, you can call either compare or process on a single invocation of the CLI.
2. That would work.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,771 @@
+             OutputStream outputStream = initOutputStream(parsedArgs, SYSTEM_OUT_DEFAULT_VALUE);
+             //Initialize error report, never return null. (default value: STDERR)
+             OutputStream errorReportOutputStream = initOutputStream(parsedArgs, SYSTEM_ERR_DEFAULT_VALUE);
+             IonWriter ionWriterForOutput = IonTextWriterBuilder.pretty().build(outputStream);
```
`--output-format` allows for formats other than text, right?

cheqianh replied:
Changed.

In case I miss anything, Ion-c uses text format for all type(text/pretty/binary) to write comparison report. Do I need to match ion-c's performance?

tgregg replied:
Yeah it's fine to use text for the comparison report.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,771 @@
+        }
+    }
+
+    private static void CompareFiles(IonWriter ionWriterForOutput,
```
Java style: method name should start with a lowercase letter.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,771 @@
+                i++;
+                j++;
+                while (!((eventsFirst.get(i).getEventType() == EventType.CONTAINER_END
+                        && eventsFirst.get(i).getIonType() == IonType.SEXP)
```
Can be SEXP or LIST.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,771 @@
+        return true;
+    }
+
+    private static boolean isSameStruct(CompareContext compareContext,
```
As we discussed offline, it should be a lot simpler to create an `IonStruct` from the sequence of events that describe the struct and then use IonEquivalence to compare them.

cheqianh replied:
Changed.

tgregg commented on test/tools/comparisonReport/ComparisonResult.java:
```diff
@@ -0,0 +1,56 @@
+            ionWriter.setFieldName("message");
+            ionWriter.writeString((message));
+        }
+        if (this.lhs != null) {
+            ionWriter.setFieldName("lhs");
+            ionWriter.stepIn(IonType.STRUCT);
+            ionWriter.setFieldName("location");
+            ionWriter.writeString(this.lhs.getLocation());
+            ionWriter.setFieldName("event");
+            this.lhs.getEvent().writeOutput(ionWriter, -1);
+            ionWriter.setFieldName("eventIndex");
+            ionWriter.writeInt(this.lhs.getEventIndex());
+            ionWriter.stepOut();
+        }
+        if (this.rhs != null) {
+            ionWriter.setFieldName("rhs");
+            ionWriter.stepIn(IonType.STRUCT);
+            ionWriter.setFieldName("location");
+            ionWriter.writeString(this.rhs.getLocation());
+            ionWriter.setFieldName("event");
+            this.rhs.getEvent().writeOutput(ionWriter, -1);
+            ionWriter.setFieldName("eventIndex");
+            ionWriter.writeInt(this.rhs.getEventIndex());
+            ionWriter.stepOut();
+        }
```
Looks like a good candidate for a method.

cheqianh replied:
Changed.

cheqianh committed:
finished cr for pr2

ready for cr

cheqianh committed:
check styles

cheqianh commented:
Thank you for the code review! I'm making small commits for each step now. I'll add a squashed commit link that includes all changes I made to make cr easier.

**code review commit**: af40cb9

**What I did for this commit:**
The above commit worked on issues from pr2. I deleted the final key word by accident, 7192d27 added it back.

**List of things I did:**

1.  Finished SYMBOL_TABLE event_type, created an append outputStream to the file.

2.  Added a space after comma.(IonProcess.java 308)

3.  Exit after finding errors of closing ionWriter. (IonProcess.java 89, 383) 

4. For issue https://github.com/cheqianh/ion-java/pull/2#discussion_r473273689, I noticed that even though the event doesn't have a field_name, ion-c will assign it $0 as long as it's in a struct. Otherwise it will throw an exception. I added this if-else statement trying to assign an event's field_name to $0 if it's in a struct without field_name , is this necessary? (ionProcess.java 432)

cheqianh committed:
finished cr issues except ionstruct and both cli

finished change to ionStruct

working on same exe

finished parse struct

finish check

ready for cr

ready for codereview

cheqianh commented:
Above commits fixed current pr3 issues except https://github.com/cheqianh/ion-java/pull/3#discussion_r473277490.
I squashed above commits into one commit  23fbf78 and force pushed it which is easier to code review.

I noticed that IonContainer doesn't have add method and adding a struct need a field_name so I think we can't simply refactor parseStruct, parseList and parseSexp.

cheqianh commented:
**Code review commit:** c1eee57
In above commit, I fixed current pr3 issues except https://github.com/cheqianh/ion-java/pull/3#discussion_r473277490 since it may create new files.
 
I noticed that IonContainer doesn't have add method and adding a struct need a field_name so I think we can't simply refactor parseStruct, parseList and parseSexp.

cheqianh commented:
**Code review commit**:  a5ec1c5

**What I did for this commit:**
In above commit, I fixed current pr3 issues except https://github.com/cheqianh/ion-java/pull/3#discussion_r473277490

I noticed that IonContainer doesn't have add method and adding a struct need a field_name so I think we can't simply write one functions for parseStruct, parseList and parseSexp.

cheqianh committed:
test

finished equivs mode

work on equivs

done equivs except embedded

finished equivs embedded

ready for cr

check styles

check styles

ready for cr

ready for cr

cheqianh commented:
Code review commit: 2720b9a

**What I did for this commit:**
1. Finished equivs mode.
Able to compare value within List ,Sexp and embedded Stream.
2. Added logic to check if a eventStream is valid (before we only checked if a single event is valid). This function is only in ionCompare file right now(Even though we will combine compare and process in the future). I'm going to apply this validation function to ionProcess as well. 
**1** check if CONTAINER_END, CONTAINER_START events are paired correctly
**2** EventStream must end up with a SREAM_END event
**3** Each string within embedded stream must end up with a STREAM_END event

**Question for ionProcess :**
Currently, I convert files to a eventStream and write outputs at the same time. Is it ok to do a 2-pass? The first pass converts files to List\<Event\> and then processing them to desired format through the second pass.

**Test:**
1. Sexp/List with multiple elements
2. Sexp/List without elements
3. embedded Stream
4. mixed of above

cheqianh committed:
working on nonequivs

almost finish non-equivs

finish non-equivs and pass tests

finished equivs-timeline and fix soem bugs

ready for cr

ready for cr

cheqianh commented:
**Code review commit** 0502356

**What I did for this commit**
This commit finished non-equivs/equivs-timeline mode.

**Description:**
1. Non-equivs mode is based on equivs mode, it used the same functions to check if the given sequence is equivalence. If the given sequence is equivs, write the report.
2. Equivs-timeline mode is based on equivs mode too. It added an extra logic in compare function. The extra code used timestamp.compareTo method  to compare SCALAR value if it's equivs-timeline mode and both values are timestamp type.

cheqianh committed:
overall check

cheqianh commented:
**code review commit:** 1e68c19

**what I did**
This commit did overall checks such as deleting some unnecessary variables, changing styles commented earlier and testing files from ion-tests.

**test:**
Tested equiv/non-eqvuis/timestamp files from https://github.com/amzn/ion-tests/tree/master/iontestdata/good

cheqianh committed:
working on refactor

little comment change

added error text and valid args

remove unnecesasry files

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+                        }
+                    }
+                } catch (IonException | NullPointerException e) {
+                    new ErrorDescription(ErrorType.STATE, e.getMessage(), path+';'+compareToPath,
```
Put a space before and after each `+`.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+            System.exit(USAGE_ERROR_EXIT_CODE);
+        }
+
+        try (//Initialize output stream, never return null. (default value: STDOUT)
```
Put a space before the `//`.

cheqianh replied:
Fixed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+
+    private static boolean compare(CompareContext compareContext,
+                                   int startI, int endI, int startJ, int endJ) throws IOException {
+        ArrayList<Event> eventsFirst = (ArrayList<Event>) compareContext.getEventStreamFirst();
+        ArrayList<Event> eventsSecond = (ArrayList<Event>) compareContext.getEventStreamSecond();
```
Why do you need to cast these to `ArrayList`?

```java
List<Event> eventsFirst = compareContext.getEventStreamFirst();
```

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+            EventType eventTypeFirst = eventFirst.getEventType();
+            EventType eventTypeSecond = eventSecond.getEventType();
+            if (eventTypeFirst != eventTypeSecond) {
+                setReportInfo(i, j,"Did't match event_type", compareContext);
```
Did't -> Didn't 

In several places.

cheqianh replied:
fixed, paste typo.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+            } else if (eventType == EventType.STREAM_END) {
+                throw new IonException("Invalid struct: eventStream ends without CONTAINER_END");
+            } else {
+                IonValue value = event.getValue();
+                value.removeFromContainer();
```
Rather than mutating the existing event, we should probably use `event.getValue().clone()`.

cheqianh replied:
Changed. (478-481) (514-517)

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+    }
+
+    private static IonList parseList(ContainerContext containerContext,
+                                               CompareContext compareContext,
+                                               int end,
+                                               boolean isFirst) {
```
Misaligned indentation.

cheqianh replied:
fixed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+        return ionList;
+    }
+
+    private static IonSexp parseSexp(ContainerContext containerContext,
```
`IonList` and `IonSexp` have a common parent: `IonSequence`. You should be able to consolidate your `parseSexp` and `parseList` methods into a single `parseSequence` method, into which you pass an `IonSequence` that was externally created from either `ION_SYSTEM.newEmptyList()` or `ION_SYSTEM.newEmptySexp()`.

cheqianh replied:
Changed. (487)

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+            } else if (eventType == EventType.STREAM_END) {
+                throw new IonException("Invalid list: eventStream ends without CONTAINER_END");
+            } else {
+                IonValue value = event.getValue();
+                value.removeFromContainer();
```
clone() instead.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+        new ComparisonResult(type,
+                new ComparisonContext(compareContext.getFile(),
+                        compareContext.getEventStreamFirst().get(compareContext.getFileEventIndex()),
+                        compareContext.getFileEventIndex() + 1),
```
Why +1?

cheqianh replied:
Changed.

Quick double check:
Does eventIndex represent the index of the event (start from 0) or the x-th event(start from 1) in the eventStream?

tgregg replied:
Index from 0.

tgregg commented on test/tools/comparisonReport/ComparisonResult.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void writeComparisonContext(IonWriter ionWriter, boolean isLeft) throws IOException {
+        ionWriter.setFieldName(isLeft ? "lhs" : "rhs");
```
Set the field name in `writeOutput`. No need to pass in a boolean and do it here.

cheqianh replied:
Changed.

tgregg commented on test/tools/comparisonReport/ComparisonResult.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        if (this.message != null) {
+            ionWriter.setFieldName("message");
+            ionWriter.writeString((message));
```
Superfluous set of parentheses.

cheqianh replied:
Changed.

tgregg commented on test/tools/comparisonReport/ComparisonResult.java:
```diff
@@ -0,0 +1,56 @@
+        ionWriter.stepIn(IonType.STRUCT);
+        if (this.result != null) {
+            ionWriter.setFieldName("result");
+            ionWriter.writeString((result.toString()));
```
Superfluous set of parentheses.

cheqianh replied:
Deleted.

tgregg commented on test/tools/comparisonReport/ComparisonResult.java:
```diff
@@ -0,0 +1,56 @@
+        ionWriter.setFieldName(isLeft ? "lhs" : "rhs");
+        ionWriter.stepIn(IonType.STRUCT);
+        ionWriter.setFieldName("location");
+        ionWriter.writeString(isLeft ? this.lhs.getLocation() : this.rhs.getLocation());
```
Make the ComparisonContext a parameter to this method and pass in either lhs or rhs.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+    private static boolean compareEquivs(CompareContext compareContext) throws IOException {
+        int i = 0;
+        int j = 0;
+        ArrayList<Event> eventStreamFirst = (ArrayList<Event>) compareContext.getEventStreamFirst();
+        ArrayList<Event> eventStreamSecond = (ArrayList<Event>) compareContext.getEventStreamSecond();
```
Use `List`.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+                throw new IonException("Embedded streams set expected.");
+            }
+
+            ArrayList<Pair<Integer, Integer>> pairsFirst;
+            ArrayList<Pair<Integer, Integer>> pairsSecond;
```
Use `List`.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+     */
+    private static ArrayList<Pair<Integer, Integer>> parseEmbeddedStream(ArrayList<Event> events, int start) {
+        ArrayList<Pair<Integer, Integer>> parsedEvents = new ArrayList<>(0);
+        int count = start + 1;
```
The name `count` is a little confusing to me. Isn't this the index of the first event in the embedded stream?

cheqianh replied:
Sorry about the confusion.

I used 'start' cursor to go through the List. For traversing each string within a embeddedStream, the string reaches the end when 'start' points to a STREAM_END event. Then We need to write the range index into a pair. start is the right index of the string and the count is for the left side.

In order to make it easier to read, I changed the 'start' to 'right', and 'count' to 'left'.
So the current logic is that using 'left', 'right' to maintain a sliding window and every time when we reach the end of a string, we write (left, right) to the list where left, right represents a string.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+     *
+     *  return List is never null. It will return a list with size zero if the embedded stream is empty.
+     */
+    private static ArrayList<Pair<Integer, Integer>> parseEmbeddedStream(ArrayList<Event> events, int start) {
```
The name of this method seems like it could be improved. Maybe something like "locateEmbeddedStreamBoundaries"?

cheqianh replied:
Sure.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+     *
+     *  return List is never null. It will return a list with size zero if the container is empty.
+     */
+    private static ArrayList<Pair<Integer, Integer>> parseContainer(ArrayList<Event> events, int start) {
```
Maybe a name like "locateContainerBoundaries" would be more descriptive?

cheqianh replied:
Sure.

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+import com.amazon.ion.system.IonReaderBuilder;
+import com.amazon.ion.system.IonSystemBuilder;
+import com.amazon.ion.util.Equivalence;
+import com.sun.tools.javac.util.Pair;
```
This import looks dangerous. Where does it come from? Is it tied to Oracle JDK?

We can just create our own Pair class if necessary.

cheqianh replied:
I created my own pair class(1237 - 1245)

tgregg commented on test/tools/cli/IonCompare.java:
```diff
@@ -0,0 +1,1004 @@
+    }
+
+    private static boolean isSameSymbolToken(SymbolToken symbolTokenX, SymbolToken symbolTokenY) {
+        if (symbolTokenX == null && symbolTokenY == null) return true;
+        else if (symbolTokenX != null && symbolTokenY != null) {
```
Style: let's always use curly braces with if/else.

cheqianh replied:
Sure, I also checked other if/else.

Question: 
Does this style apply for if/else branch only or for all single if statement as well?

tgregg replied:
That's what I recommend.

tgregg commented:
> Currently, I convert files to a eventStream and write outputs at the same time. Is it ok to do a 2-pass? The first pass converts files to List<Event> and then processing them to desired format through the second pass.

Yes, two-pass is a good idea.

tgregg commented:
> For issue #2 (comment), I noticed that even though the event doesn't have a field_name, ion-c will assign it $0 as long as it's in a struct. Otherwise it will throw an exception. I added this if-else statement trying to assign an event's field_name to $0 if it's in a struct without field_name , is this necessary? (ionProcess.java 432)

Can you give me an example of this? I'm not sure I fully understand.

cheqianh committed:
finished basic issues, ready to work on refactor ionsequnces

finished refactor and work on pairs

created pair strcut and delete untrusted pair

ready for cr

cheqianh commented:
> > For issue #2 (comment), I noticed that even though the event doesn't have a field_name, ion-c will assign it $0 as long as it's in a struct. Otherwise it will throw an exception. I added this if-else statement trying to assign an event's field_name to $0 if it's in a struct without field_name , is this necessary? (ionProcess.java 432)
> 
> Can you give me an example of this? I'm not sure I fully understand.

Sure,  For event stream:
```
$ion_event_stream
{
  event_type: CONTAINER_START,
  ion_type: STRUCT,
  depth: 0
}
{
  event_type: SCALAR,
  ion_type: INT,
  value_text: "1",
 // here is supposed to have a field_name
  value_binary: [
    224,
    1,
    0,
    234,
    33,
    1
  ],
  depth: 1
}
{
  event_type: CONTAINER_END,
  ion_type: STRUCT,
  depth: 0
}
{
  event_type: STREAM_END,
  depth: 0
}
```

If I use ion-c to process this eventstream to ionstream, it printed:
```
{
  $0: 1
}
```

Event though the event stream doesn't have a field_name, (which might be an error) it outputs $0 instead.

cheqianh commented:
**Code review commit:** 54b458a

**What I did**
This commit fix the issue https://github.com/cheqianh/ion-java/pull/3#discussion_r473277490. This commit just refactor  functions but didn't fix any issues above. The next commit fixed issues above.

For this commit, file IonProcess.java and ionCompare.java is not necessary to code review anymore. I'll deleted them later.
Refactor doesn't affact equivs/non-equivs/eqvuis-timeline commits since they used same logic.

cheqianh commented:
code review commit: e14795c

**What I did:**
This commit fixes the current code review issues.(which you just code reviewed) I didn't delete ionProcess and ionCompare right now in case there are something important. I'll delete them after finish current issues.

**Question**
I tried to write output directly to system.out when a symbol_table event faced but it didn't work.
FileOutputStream append works if the outputStream is a file. If the outputStream is system.out, Is printStream append method a good idea to do the same thing as FileOutputStream? It seems like printStream.append only support charsequence which might not be correct.

tgregg commented:
> Event though the event stream doesn't have a field_name, (which might be an error) it outputs $0 instead.

Yeah, that looks like a bug in the ion-c implementation. That should be an error. Can you please open an issue in ion-c?

tgregg commented on test/tools/cli/IonJavaCli.java:
```diff
@@ -0,0 +1,1341 @@
+                    IonWriter ionWriterForOutput = outputFormat.createIonWriter(outputStream);
+                    IonWriter ionWriterForErrorReport = outputFormat.createIonWriter(errorReportOutputStream);
+            ) {
+                compareFiles(ionWriterForOutput, ionWriterForErrorReport, parsedArgs, comparisonType);
```
I would move the `commandType` check inside this `try` so you can reuse a single `try` block. No need to duplicate it in a branch for the other command type.

tgregg commented on test/tools/cli/IonJavaCli.java:
```diff
@@ -0,0 +1,1341 @@
+                throw new IonException("Invalid struct: eventStream ends without CONTAINER_END");
+            } else {
+                IonValue cloneValue = event.getValue().clone();
+                cloneValue.removeFromContainer();
```
I believe cloning already removes the value from its container, so this line is probably redundant.

tgregg commented on test/tools/cli/IonJavaCli.java:
```diff
@@ -0,0 +1,1341 @@
+                throw new IonException("Invalid ionSequence: eventStream ends without CONTAINER_END");
+            } else {
+                IonValue cloneValue = event.getValue().clone();
+                cloneValue.removeFromContainer();
```
I believe cloning already removes the value from its container, so this line is probably redundant.

tgregg commented on test/tools/cli/IonJavaCli.java:
```diff
@@ -0,0 +1,1341 @@
+    }
+
+    static class Pair {
+        final int fst;
+        final int snd;
```
Now that we have control over this class, let's give these better names. Maybe `left` and `right` to match the variable names you use to construct them.

tgregg reviewed (approved):
My remaining comments are pretty minor, so you can fix them and then merge to close out this PR.

tgregg commented:
> I tried to write output directly to system.out when a symbol_table event faced but it didn't work.
FileOutputStream append works if the outputStream is a file. If the outputStream is system.out, Is printStream append method a good idea to do the same thing as FileOutputStream? It seems like printStream.append only support charsequence which might not be correct.

I believe the problem is that closing the `IonWriter` closes the underlying `OutputStream`, and once `System.out` is closed, continuing to write to it will not cause the output to appear in stdout. To avoid this, you could create a custom trivial OutputStream implementation that delegates to System.out for everything except `close()`. That way closing the `IonWriter` will not close `System.out` and you can continue writing to it. Take a look at [FilterOutputStream](https://docs.oracle.com/javase/7/docs/api/java/io/FilterOutputStream.html).

cheqianh committed:
deleted 2 files and small change

fixing bugs

rename pair left/right

fixed issues

cheqianh commented:
above commits fixed issues rather than SYMBAL_TABLE/System.out one

cheqianh committed:
fix system.out symbol_table issue

cheqianh commented:
commits 0306b96 added some logic to make sure system.out isn't closed when SYMBOL_TABLE event faced.

cheqianh commented on src/com/amazon/tools/cli/IonJavaCli.java:
```diff
@@ -1205,9 +1206,7 @@ private static BufferedOutputStream initOutputStream(CommandArgs args,
                     outputStream = new BufferedOutputStream(out, BUFFER_SIZE);
                 } else {
-                    OutputStream out = new FilterOutputStream(System.out);
-                    out.close();
```
These 2 lines are debug log, Already deleted.

cheqianh commented on src/com/amazon/tools/cli/NoCloseOutputStream.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public void close() throws IOException{
+        out.flush();
```
To unable the close method for system.out, Do I need to out.flush here?
